diff --git a/dist/chunk-tasks.js b/dist/chunk-tasks.js
index 8d5d39d8ecfa64017593b50327215c95b600467a..8fb7f0add58b4aa6f173a8b1655498bf401ab0a8 100644
--- a/dist/chunk-tasks.js
+++ b/dist/chunk-tasks.js
@@ -1,4 +1,3 @@
-import { processError } from '@vitest/utils/error';
 import { relative } from 'pathe';
 import { toArray } from '@vitest/utils';
 
@@ -103,14 +102,18 @@ function checkAllowOnly(task, allowOnly) {
   if (allowOnly) {
     return;
   }
-  const error = processError(
-    new Error(
-      "[Vitest] Unexpected .only modifier. Remove it or pass --allowOnly argument to bypass this error"
-    )
-  );
+  const error = new Error(
+    "[Vitest] Unexpected .only modifier. Remove it or pass --allowOnly argument to bypass this error"
+  )
   task.result = {
     state: "fail",
-    errors: [error]
+    errors: [
+      {
+        name: error.name,
+        stack: error.stack,
+        message: error.message
+      }
+    ]
   };
 }
 function generateHash(str) {
